c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,400 @@
+    ) -> Option<(
+        Location,
+        mir::Place<'tcx>,
+        &'a mir::Operand<'tcx>,
+        &'a [mir::Operand<'tcx>],
+    )> {
+        let loc = *locs.last()?;
+        let term = self.mir.stmt_at(loc).right()?;
+        match term.kind {
+            mir::TerminatorKind::Call {
+                ref func,
+                ref args,
+                destination,
+                ..
+            } => Some((loc, destination, func, args)),
+            _ => None,
+        }
+    }
+
+    fn should_ignore_statement(&self, loc: Location) -> bool {
+        if let Some(stmt) = self.mir.stmt_at(loc).left() {
+            match stmt.kind {
+                mir::StatementKind::FakeRead(..)
+                | mir::StatementKind::StorageLive(..)
+                | mir::StatementKind::StorageDead(..)
+                | mir::StatementKind::Nop => return true,
+                _ => {}
+            }
+        }
+        false
+    }
+
+    fn visit_expr_inner(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        let _g = panic_detail::set_current_span(ex.span);
+
+        let locs = self
+            .span_index
+            .lookup_exact(ex.span)
+            .copied()
+            .filter(|&loc| !self.should_ignore_statement(loc))
+            .collect::<Vec<_>>();
+        if locs.is_empty() {
+            return;
+        }
+
+        let warn = |desc| {
+            warn!("{}", desc);
+            info!("locs:");
+            for &loc in &locs {
+                self.mir.stmt_at(loc).either(
+                    |stmt| info!("  {:?}: {:?}", locs, stmt),
+                    |term| info!("  {:?}: {:?}", locs, term),
+                );
+            }
+            info!("span = {:?}", ex.span);
+            info!("expr = {:?}", ex);
+        };
+
+        // Most exprs end with an assignment, storing the result into a temporary.
+        match ex.kind {
+            hir::ExprKind::Assign(pl, rv, _span) => {
+                // For `Assign`, we expect the assignment to be the whole thing.
+                let (loc, _mir_pl, _mir_rv) = match self.get_sole_assign(&locs) {
+                    Some(x) => x,
+                    None => {
+                        warn("expected exactly one StatementKind::Assign");
+                        return;
+                    }
+                };
+                self.record(loc, &[], ex);
+                self.record(loc, &[SubLoc::Dest], pl);
+                self.record(loc, &[SubLoc::Rvalue], rv);
+            }
+
+            hir::ExprKind::Call(_, args) | hir::ExprKind::MethodCall(_, args, _) => {
+                let (loc, _mir_pl, _mir_func, mir_args) = match self.get_last_call(&locs) {
+                    Some(x @ (_, pl, _, _)) if is_var(pl) => x,
+                    _ => {
+                        warn("expected final Call to store into var");
+                        return;
+                    }
+                };
+                self.record_desc(loc, &[], ex, MirOriginDesc::StoreIntoLocal);
```

Comment:
why is there a distinction between `StoreIntoLocal` and `Expr` with a `Dest`, i.e. why not `self.record(loc, &[SubLoc::Dest], mir_pl);`?

Reply:
Ideally there would be only one `MirOriginDesc::Expr` entry for each HIR expr, indicating the piece of MIR that's primarily responsible for implementing the evaluation of that expr.  For `x + y` this should be an `Rvalue::BinaryOp`, for `f(x)` it's a `TerminatorKind::Call`, etc.  For a store into a local like `_z = _x + _y`, it would be confusing to label both the `Place` (`SubLoc::Dest`) and the `Rvalue` as being "the expr itself".

Reply:
I think I understand. Plus, I'm not sure there is an `hir::Expr` that represents the destination of the method call anyway, right?

Reply:
Right - the LHS in the MIR is a temporary, so there is no corresponding variable in the HIR code.